### PR TITLE
refactor using context managers classes for resources

### PR DESCRIPTION
**Pull request by lesnik512**

### CODE REVIEW

**Comment by lesnik512** [comment]
@zerlok @alexanderlazarev0 Hi) So here is my version with only allowed generators (sync and async) and context manager classes.

**Comment by zerlok** [comment]
@lesnik512 , why did you decide to remove `Callable[P, AsyncContextManager[T]]` support ?

**Comment by lesnik512** [comment]
Because it can fail in runtime and require quite dirty hacks. And it doesn't seem very useful because instead of using wrapped generators one always can use not wrapped generator

**Comment by zerlok** [comment]
@lesnik512 , let's have a deeper look on what can fail during runtime? Which hacks bothers you? I believe we can find a solution. 

I didn't get your point about wrapped / not wrapped generator. Why do you think it's not useful to make a factory method inside a class that can provide a resource asynchronously (via async context manager interface)? Why it's not useful to reuse that method in the DI container? Actually, that was the main case for my proposal https://github.com/modern-python/that-depends/issues/79#issuecomment-2455062784

**Comment by lesnik512** [comment]
Ok)
Let's start from the original proposal
This one:
```python
class MySender:
    @asynccontextmanager
    @classmethod
    async def create(cls, conn: AMQPConnection) -> typing.AsyncIterator[MyPublisher]:
        exchange = await conn.declare_exchange(...)
        try:
            yield cls(exchange)
        finally:
            await conn.delete_exchange(exchange)

    def __init__(self, exchange: AMQPExchange) -> None:
        self.__exchange = exchange

    async def send(self, msg: bytes) -> None:
        await self.__exchange.publish_message(msg)
```
is almost the same as:
```
class MySender:
    @classmethod
    async def create(cls, conn: AMQPConnection) -> typing.AsyncIterator[MyPublisher]:
        exchange = await conn.declare_exchange(...)
        try:
            yield cls(exchange)
        finally:
            await conn.delete_exchange(exchange)

    def __init__(self, exchange: AMQPExchange) -> None:
        self.__exchange = exchange

    async def send(self, msg: bytes) -> None:
        await self.__exchange.publish_message(msg)
```
And the second option already can be used. So why do we need to accept decorated generator if we can use generator without decorator?

**Comment by lesnik512** [comment]
About failing in runtime: my bad. I misunderstood this part https://github.com/modern-python/that-depends/pull/113/files#diff-e69ae086d80699587fd1eba7cff16558f6f3e145e7a70f288577cbd691e4c441R210

But still it doesn't feel right to have TypeError, when resolving dependency

**Comment by lesnik512** [comment]
In common, it seems to me like allowing decorated generators to resources adds too much complexity with not very clear benefits.

**Comment by lesnik512** [comment]
My bad, that I didn't pay enough attention to the original issue and haven't provide my concerns beforehand. I thought, the main idea was to allow to pass subclasses of typing.ContextManager and typing.AsyncContextManager.

Sorry, man

**Comment by zerlok** [comment]
@lesnik512 , it's ok, I'm glad we can sort things out :) 

1) I chose TypeError, since it was type check isinstance . class supports both CMs (sync & async CM), but only `sync_resolve`  function can't support async CM, because it can't invoke async code. So only one method needs a specific type and cant support another. 

From python 3.12 chapter about exceptions:

TypeError - Raised when an operation or function is applied to an object of inappropriate type.

RuntimeError - Raised when an error is detected that doesn’t fall in any of the other categories.

RuntimeError feels like an appropriate exception here, but I'm not sure it is. From my point of view the ideal solution is to create a custom exception type for Resource (like ResourceError / ResourceResolvingError) and raise it instead. It's much cleaner and easy to use & support.

2) For proposal. The `MySender` code from the [snippet](https://github.com/modern-python/that-depends/pull/115#issuecomment-2466818781) is a part of application module that can be reused in other code. E.g. I can write in other client code 

```python
async def do_some_stuff():
    async with MySender.create(...) as my_sender:  # use created sender via async CM
        await my_sender.send(...)
```

If I apply your 2nd suggestion [from here](https://github.com/modern-python/that-depends/pull/115#issuecomment-2466818781) then I have to add one more method or add decorator in the client code e.g.:

```python
async def do_some_stuff():
    async with asynccontextmanager(MySender.create(...)) as my_sender:  # use created sender via async CM
        await my_sender.send(...)
```

That seems like overhead to me.

So, basically this code (classmethod + asynccontextmanager decorator) provides an ability to write async constructors in python. The returned value is a async CM which can be automatically cleaned up (on context exit). The class method may require "setup" arguments and those arguments may come from the client code (inline in async with statement) or from DI providers.

**Comment by zerlok** [comment]
The asynccontextmanager is a decorator from builtins. It is meant to be used as decorator on functions. So I am pretty sure that  is a common practice and it would be great to support it too.

```python
@asynccontextmanager
async def provide_stuff(...) -> AsyncIterator[...]: ...

reveal_type(provide_stuff) # Callable[P, AsyncContextManager[V_co]]
```

**Comment by lesnik512** [comment]
1. About TypeError. My idea is not to create 100500+ custom exceptions in cases, when there is no reason to caught them. Most of the exceptions in the library are the signal of misuse and require to change something.
2. Instead of this:
```python
async def do_some_stuff():
    async with asynccontextmanager(MySender.create(...)) as my_sender:  # use created sender via async CM
        await my_sender.send(...)
```
You can write something like:
```python
class MySender:
    @classmethod
    async def create(cls, conn: AMQPConnection) -> typing.AsyncIterator[MyPublisher]:
        exchange = await conn.declare_exchange(...)
        try:
            yield cls(exchange)
        finally:
            await conn.delete_exchange(exchange)

    create_cm = asynccontextmanager(create)

    def __init__(self, exchange: AMQPExchange) -> None:
        self.__exchange = exchange

    async def send(self, msg: bytes) -> None:
        await self.__exchange.publish_message(msg)
```

And you can use create_cm as a context manager where do you need

**Comment by lesnik512** [comment]
> The asynccontextmanager is a decorator from builtins. It is meant to be used as decorator on functions. So I am pretty sure that is a common practice and it would be great to support it too.

But it doesn't mean that we must accept everything.

I think, this part of python's zen can help to express my opinion on this:

> There should be one-- and preferably only one --obvious way to do it.

**Comment by zerlok** [comment]
From my perspective, you are trying to use one way, yes. But it is not the way how devs use it in the code. It is usually `@asynccontextmanager` decorator to a function, not an iterable.

**Comment by lesnik512** [comment]
It is used this way in dependency-injector and this is the simplest way.

For more advanced cases you can use classes which implement abstract classes from typing module: ContextManager and AsyncContextManager.

We are walking circles now) Your case can be covered by current implementation and I still see no reason to add decorated generators to Resources

**Comment by zerlok** [comment]
> no reason to add decorated generators to Resources

It's not about decorated generators. It's about factory CM . Callable[P, AsyncContextManager] . CM object can't be parametrized. That's why you need some factory function for CM.

**Comment by lesnik512** [comment]
I see no difference between decorated generators and Callable[P, AsyncContextManager]

This one I'm calling decorated generator:
```python
    @asynccontextmanager
    @classmethod
    async def create(cls, conn: AMQPConnection) -> typing.AsyncIterator[MyPublisher]:
        exchange = await conn.declare_exchange(...)
        try:
            yield cls(exchange)
        finally:
            await conn.delete_exchange(exchange)
```

**Comment by lesnik512** [comment]
By the way this one is also Callable[P, AsyncContextManager] and if you add init method, it can be parametrised
```python
class ContextManagerResource(typing.ContextManager[datetime.datetime]):
    def __enter__(self) -> datetime.datetime:
        logger.debug("Resource initiated")
        return datetime.datetime.now(tz=datetime.timezone.utc)

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_value: BaseException | None,
        traceback: types.TracebackType | None,
    ) -> None:
        logger.debug("Resource destructed")
```

**Comment by zerlok** [comment]
Just checked the latest that-depends version 1.24.0 , but the following code fails

```python
import asyncio
import typing as t

from protomq.abc import BinaryPublisher
from protomq.broker import Broker
from that_depends import BaseContainer, providers


def _get_pub(broker: Broker) -> t.AsyncContextManager[BinaryPublisher]:
    return broker.publisher()


# reveal_type(_get_pub)


class DI(BaseContainer):
    broker = providers.Resource(Broker, dsn="amqp://guest:guest@localhost:5672/")
    pub = providers.Resource(_get_pub, broker.cast)


async def main() -> None:
    try:
        await DI.init_resources()
        print(await DI.pub())

    finally:
        await DI.tear_down()


if __name__ == "__main__":
    asyncio.run(main())
```

```
TypeError: Unsupported resource type
```

mypy:

```
examples/cc.py:13: note: Revealed type is "def (broker: protomq.broker.Broker) -> typing.AsyncContextManager[protomq.abc.Publisher[protomq.message.Message[builtins.bytes], Union[builtins.bool, None]]]"
Success: no issues found in 1 source file
```

**Comment by lesnik512** [comment]
> Just checked the latest that-depends version 1.24.0 , but the following code fails

Why wouldn't?) Now only following is accepted by Resource provider:
```python
import datetime
import logging
import types
import typing


logger = logging.getLogger(__name__)


async def create_async_resource() -> typing.AsyncIterator[datetime.datetime]:
    logger.debug("Async resource initiated")
    try:
        yield datetime.datetime.now(tz=datetime.timezone.utc)
    finally:
        logger.debug("Async resource destructed")


def create_sync_resource() -> typing.Iterator[datetime.datetime]:
    logger.debug("Resource initiated")
    try:
        yield datetime.datetime.now(tz=datetime.timezone.utc)
    finally:
        logger.debug("Resource destructed")


class ContextManagerResource(typing.ContextManager[datetime.datetime]):
    def __int__(self):
        pass
    
    def __enter__(self) -> datetime.datetime:
        logger.debug("Resource initiated")
        return datetime.datetime.now(tz=datetime.timezone.utc)

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_value: BaseException | None,
        traceback: types.TracebackType | None,
    ) -> None:
        logger.debug("Resource destructed")


class AsyncContextManagerResource(typing.AsyncContextManager[datetime.datetime]):
    def __int__(self):
        pass
    
    async def __aenter__(self) -> datetime.datetime:
        logger.debug("Async resource initiated")
        return datetime.datetime.now(tz=datetime.timezone.utc)

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_value: BaseException | None,
        traceback: types.TracebackType | None,
    ) -> None:
        logger.debug("Async resource destructed")
```

Your snippet can be rewritten sth like this:
```python
@asynccontextmanager
async def _get_pub(broker: Broker) -> t.AsyncIterator[BinaryPublisher]:
    async with broker.publisher() as publisher
        yield publisher
```

**Comment by zerlok** [comment]
@lesnik512 

Please, have a look at mypy type check from my [previous message](https://github.com/modern-python/that-depends/pull/115#issuecomment-2466876504). Mypy said - it is ok to pass `_get_pub` with decorated `@asynccontextmanager`  to Resource. But the program failed during the runtime. Looks like a bug.

I completely disagree to use AsyncIterator function as creator main type in resource, because its signature is not for one time created instances. Such signature is not for `async with` statements either. AsyncIterator is designed to be iterated unbounded amount of times (could be 1, could be 2, could be 100500 or even 0). The idea to allow to use it as CM in the specific cases is a hack that lowers the amount of code needed to be written by developer to create a CM. That's why `asynccontextmanager` decorator was created for. It guarantees the specific case of AsyncIterator function, which should yield result only once .

In my opinion the current Resource creator type is too broad and can not be completely supported and has false-positives during type checking. Client can't be sure that Resource will work properly with his AsyncIterator function during the runtime, because client doesn't know that function should yield result only once, an client can't check it with static analyzers. When client is forced to pass AsyncContextManager it is more straightforward to him how to design a valid creator.

So, I'm asking you to reconsider the usage of Resource .

**Comment by lesnik512** [comment]
I do not think, that type checking must catch all negative cases. Something can be learned from docs and examples and checked on runtime.

It was done this way in dependency-injector and this is a time-tested option and I prefer to leave it this way.

Your proposal is adding unnecessary complexity with too little benefit for special cases, so I stand by my decision, because I am responsible for ensuring that the package remains maintainable.
